xwiki-platform-core/xwiki-platform-security/xwiki-platform-security-authorization/xwiki-platform-security-authorization-api/src/main/java/org/xwiki/security/authorization/internal/AbstractAuthorizationSettler.java

```diff
@@ -245,6 +246,37 @@ protected XWikiSecurityAccess applyDefaults(UserSecurityReference user, Security
         return access;
     }
 
+    /**
+     * Deny all rights that are lacking needed rights.
+     *
+     * @param access the accumulated access result (modified and returned)
+     * @return the accumulated access result
+     * @since 15.10.2
+     */
+    protected XWikiSecurityAccess applyNeededRights(XWikiSecurityAccess access)
+    {
+        // Check needed rights.
+        boolean rightsModified;
+        do {
+            rightsModified = false;
+            for (Right right : Right.values()) {
+                // For every right that is allowed, check if all needed rights are allowed.
+                if (access.get(right) == RuleState.ALLOW) {
+                    for (Right neededRight : right.getNeededRights()) {
+                        if (access.get(neededRight) != RuleState.ALLOW) {
+                            access.set(right, RuleState.DENY);
```

Comment:
What about adding a debug log here to help in case of problem?

Reply:
I'm also wondering if there shouldn't be an hardcoded mechanism for PR right. Let's imagine that for some reason PR right is configured to need VIEW. Right now, you completely lock down a wiki from anyone including PR admins, with this loop.